File: sdks/elixir/lib/kyuubiki_sdk/model_research_frontier.ex
defmodule KyuubikiSdk.ModelResearchFrontier do
  @moduledoc "Verified cross-turn progression for model-assisted research jobs."

  alias KyuubikiSdk.Error
  alias KyuubikiSdk.ModelCollaboration
  alias KyuubikiSdk.ModelPlanApproval
  alias KyuubikiSdk.ModelResearchExecution

  @frontier_schema "kyuubiki.model-research-frontier/v2"
  @submit_actions ["fem_submit", "workflow_submit_catalog", "workflow_submit_graph"]

  def schema_version, do: @frontier_schema

  def compute_digest(current) when is_map(current) do
    with :ok <- validate(current) do
      current
      |> digest_projection()
      |> ModelPlanApproval.compute_canonical_digest()
    end
  end

  def compute_digest(_current), do: validation_error("research frontier must be a JSON object")

  def verify_digest(current, expected_digest) do
    cond do
      not valid_plan_digest?(expected_digest) ->
        validation_error("expected research frontier digest is invalid")

      true ->
        with {:ok, actual_digest} <- compute_digest(current) do
          if actual_digest == expected_digest,
            do: :ok,
            else:
              validation_error("persisted research frontier digest does not match trusted state")
        end
    end
  end

  def digest_verifier(expected_digest) do
    if valid_plan_digest?(expected_digest) do
      {:ok, fn current -> verify_digest(current, expected_digest) end}
    else
      validation_error("expected research frontier digest is invalid")
    end
  end

  def start(receipt, receipt_verifier) when is_function(receipt_verifier, 1) do
    with :ok <- validate_receipt(receipt),
         :ok <- verify_receipt(receipt, receipt_verifier),
         {:ok, record} <- last_record(receipt) do
      start_from_record(receipt, record)
    end
  end

  def start(_receipt, _receipt_verifier),
    do: validation_error("receipt verifier must be a one-argument function")

  def advance(current, receipt, frontier_verifier, receipt_verifier)
      when is_function(frontier_verifier, 1) and is_function(receipt_verifier, 1) do
    with :ok <- validate(current),
         :ok <- verify_frontier(current, frontier_verifier),
         :ok <- validate_receipt(receipt),
         :ok <- verify_receipt(receipt, receipt_verifier),
         :ok <- validate_identity(current, receipt),
         {:ok, expected} <- next_action(current),
         {:ok, record} <- last_record(receipt) do
      advance_from_record(current, receipt, record, expected)
    end
  end

  def advance(_current, _receipt, _frontier_verifier, _receipt_verifier),
    do: validation_error("frontier and receipt verifiers must be one-argument functions")

  def build_proposal(current, frontier_verifier) when is_function(frontier_verifier, 1) do
    with :ok <- validate(current),
         :ok <- verify_frontier(current, frontier_verifier),
         {:ok, action} <- next_action(current),
         {:ok, job_id} <- required_string(current, "job_id") do
      {:ok,
       %{
         "schema_version" => ModelCollaboration.proposal_schema_version(),
         "session_id" => current["session_id"],
         "summary" => "Advance verified research frontier with #{action}",
         "calls" => [
           %{
             "id" => "frontier-#{current["transition_count"] + 1}-#{action}",
             "action" => action,
             "payload" => %{"job_id" => job_id},
             "reason" => "Use the job identifier retained from verified execution evidence"
           }
         ]
       }}
    end
  end

  def build_proposal(_current, _frontier_verifier),
    do: validation_error("frontier verifier must be a one-argument function")

  defp start_from_record(%{"status" => "failed"} = receipt, record),
    do: {:ok, blocked_frontier(receipt, record, receipt["plan_digest"], nil, 1)}

  defp start_from_record(%{"status" => "completed"} = receipt, record) do
    cond do
      record["action"] not in @submit_actions ->
        validation_error("initial research receipt must end with a supported job submission")

      true ->
        case get_in(record, ["output", "job", "job_id"]) do
          job_id when is_binary(job_id) and job_id != "" ->
            {:ok,
             frontier(receipt, record,
               origin_plan_digest: receipt["plan_digest"],
               stage: "waiting_for_job",
               job_id: job_id,
               next_action: "job_wait",
               transition_count: 1
             )}

          _ ->
            validation_error("job submission receipt did not contain job.job_id")
        end
    end
  end

  defp advance_from_record(current, %{"status" => "failed"} = receipt, record, _expected) do
    {:ok,
     blocked_frontier(
       receipt,
       record,
       current["origin_plan_digest"],
       current["job_id"],
       current["transition_count"] + 1
     )}
  end

  defp advance_from_record(current, receipt, record, expected) do
    cond do
      record["action"] != expected ->
        validation_error(
          "research receipt ended with #{inspect(record["action"])}; frontier requires #{expected}"
        )

      record["job_id"] != current["job_id"] ->
        validation_error("research receipt job_id does not match frontier binding")

      expected == "job_wait" ->
        advance_wait(current, receipt, record)

      expected == "result_fetch" ->
        {:ok,
         frontier(receipt, record,
           origin_plan_digest: current["origin_plan_digest"],
           stage: "ready_to_validate",
           job_id: current["job_id"],
           next_action: nil,
           transition_count: current["transition_count"] + 1
         )}

      true ->
        validation_error("unsupported frontier next action: #{expected}")
    end
  end

  defp advance_wait(current, receipt, record) do
    case get_in(record, ["output", "terminal", "job", "status"]) do
      "completed" = status ->
        {:ok,
         frontier(receipt, record,
           origin_plan_digest: current["origin_plan_digest"],
           stage: "ready_to_fetch_result",
           job_id: current["job_id"],
           next_action: "result_fetch",
           transition_count: current["transition_count"] + 1,
           job_status: status
         )}

      status when status in ["failed", "cancelled"] ->
        {:ok,
         frontier(receipt, record,
           origin_plan_digest: current["origin_plan_digest"],
           stage: "blocked",
           job_id: current["job_id"],
           next_action: nil,
           transition_count: current["transition_count"] + 1,
           job_status: status,
           blocking_reason: "job reached terminal status #{status}"
         )}

      nil ->
        validation_error("job_wait receipt did not contain terminal.job.status")

      status ->
        validation_error("job_wait returned non-terminal status #{inspect(status)}")
    end
  end

  defp frontier(receipt, record, opts) do
    %{
      "schema_version" => @frontier_schema,
      "session_id" => receipt["session_id"],
      "workflow_id" => receipt["workflow_id"],
      "origin_plan_digest" => Keyword.fetch!(opts, :origin_plan_digest),
      "stage" => Keyword.fetch!(opts, :stage),
      "job_id" => Keyword.get(opts, :job_id),
      "next_action" => Keyword.get(opts, :next_action),
      "transition_count" => Keyword.fetch!(opts, :transition_count),
      "evidence" => %{
        "approval_id" => receipt["approval_id"],
        "plan_digest" => receipt["plan_digest"],
        "action" => record["action"],
        "record_index" => record["index"],
        "authority" => record["authority"],
        "job_status" => Keyword.get(opts, :job_status)
      },
      "blocking_reason" => Keyword.get(opts, :blocking_reason)
    }
  end

  defp blocked_frontier(receipt, record, origin_plan_digest, job_id, transition_count) do
    frontier(receipt, record,
      origin_plan_digest: origin_plan_digest,
      stage: "blocked",
      job_id: job_id,
      next_action: nil,
      transition_count: transition_count,
      blocking_reason: record["error"] || "research execution failed"
    )
  end

  defp validate_receipt(receipt) when is_map(receipt) do
    cond do
      receipt["schema_version"] != ModelResearchExecution.receipt_schema_version() or
          receipt["execution_authority"] != "kyuubiki-headless-sdk" ->
        validation_error("unsupported or untrusted research execution receipt")

      not present_string?(receipt["session_id"]) or
        not present_string?(receipt["workflow_id"]) or
        not valid_plan_digest?(receipt["plan_digest"]) or
        not is_list(receipt["records"]) or receipt["records"] == [] ->
        validation_error("research execution receipt is incomplete")

      receipt["status"] not in ["completed", "failed"] ->
        validation_error("research execution receipt has an unsupported status")

      not valid_final_record?(receipt["status"], List.last(receipt["records"])) ->
        validation_error("research execution receipt has an invalid final record")

      true ->
        :ok
    end
  end

  defp validate_receipt(_receipt),
    do: validation_error("research execution receipt must be a JSON object")

  def validate(current) when is_map(current) do
    cond do
      current["schema_version"] != @frontier_schema or
        not present_string?(current["session_id"]) or
        not present_string?(current["workflow_id"]) or
        not valid_plan_digest?(current["origin_plan_digest"]) or
        not is_map(current["evidence"]) or
        not valid_plan_digest?(current["evidence"]["plan_digest"]) or
        not valid_evidence?(current["evidence"]) or
        not is_integer(current["transition_count"]) or current["transition_count"] <= 0 ->
        validation_error("research frontier is incomplete or uses an unsupported schema")

      not valid_frontier_state?(current) ->
        validation_error("research frontier stage and next action are inconsistent")

      true ->
        :ok
    end
  end

  def validate(_current),
    do: validation_error("research frontier must be a JSON object")

  defp verify_receipt(receipt, verifier) do
    case verifier.(receipt) do
      true ->
        :ok

      :ok ->
        :ok

      {:ok, _evidence} ->
        :ok

      {:error, reason} ->
        validation_error("caller receipt verifier rejected receipt: #{inspect(reason)}")

      _ ->
        validation_error("caller receipt verifier rejected research execution receipt")
    end
  end

  defp verify_frontier(current, verifier) do
    case verifier.(current) do
      true ->
        :ok

      :ok ->
        :ok

      {:ok, _evidence} ->
        :ok

      {:error, reason} ->
        validation_error("caller frontier verifier rejected frontier: #{inspect(reason)}")

      _ ->
        validation_error("caller frontier verifier rejected research frontier")
    end
  end

  defp validate_identity(current, receipt) do
    if receipt["session_id"] == current["session_id"] and
         receipt["workflow_id"] == current["workflow_id"],
       do: :ok,
       else: validation_error("research receipt does not match frontier session and workflow")
  end

  defp next_action(current) do
    case current["next_action"] do
      value when is_binary(value) and value != "" -> {:ok, value}
      _ -> validation_error("research frontier has no executable next action")
    end
  end

  defp required_string(value, key) do
    case value[key] do
      item when is_binary(item) and item != "" -> {:ok, item}
      _ -> validation_error("research frontier has no bound #{key}")
    end
  end

  defp last_record(receipt) do
    case List.last(receipt["records"]) do
      record when is_map(record) -> {:ok, record}
      _ -> validation_error("research execution receipt has an invalid final record")
    end
  end

  defp present_string?(value), do: is_binary(value) and String.trim(value) != ""

  defp valid_plan_digest?("sha256:" <> digest) when byte_size(digest) == 64,
    do: String.match?(digest, ~r/\A[0-9a-f]{64}\z/)

  defp valid_plan_digest?(_value), do: false

  defp valid_evidence?(evidence) do
    is_binary(evidence["action"]) and
      Regex.match?(~r/\A[a-z][a-z0-9_]*\z/, evidence["action"]) and
      is_integer(evidence["record_index"]) and evidence["record_index"] > 0 and
      (is_nil(evidence["approval_id"]) or is_binary(evidence["approval_id"])) and
      (is_nil(evidence["authority"]) or is_binary(evidence["authority"])) and
      evidence["job_status"] in [nil, "completed", "failed", "cancelled"]
  end

  defp digest_projection(current) do
    evidence = current["evidence"]

    %{
      "schema_version" => current["schema_version"],
      "session_id" => current["session_id"],
      "workflow_id" => current["workflow_id"],
      "origin_plan_digest" => current["origin_plan_digest"],
      "stage" => current["stage"],
      "job_id" => current["job_id"],
      "next_action" => current["next_action"],
      "transition_count" => current["transition_count"],
      "evidence" => %{
        "approval_id" => evidence["approval_id"],
        "plan_digest" => evidence["plan_digest"],
        "action" => evidence["action"],
        "record_index" => evidence["record_index"],
        "authority" => evidence["authority"],
        "job_status" => evidence["job_status"]
      },
      "blocking_reason" => current["blocking_reason"]
    }
  end

  defp valid_final_record?("completed", record) when is_map(record),
    do:
      is_nil(record["error"]) and not is_nil(record["output"]) and is_binary(record["authority"])

  defp valid_final_record?("failed", record) when is_map(record),
    do: is_binary(record["error"])

  defp valid_final_record?(_status, _record), do: false

  defp valid_frontier_state?(current) do
    has_job = present_string?(current["job_id"])

    case current["stage"] do
      "waiting_for_job" ->
        has_job and current["next_action"] == "job_wait" and is_nil(current["blocking_reason"])

      "ready_to_fetch_result" ->
        has_job and current["next_action"] == "result_fetch" and
          is_nil(current["blocking_reason"])

      "ready_to_validate" ->
        has_job and is_nil(current["next_action"]) and is_nil(current["blocking_reason"])

      "blocked" ->
        is_nil(current["next_action"]) and present_string?(current["blocking_reason"])

      _ ->
        false
    end
  end

  defp validation_error(message), do: {:error, Error.model_research_execution([message])}
end

File: sdks/python/kyuubiki_sdk/model_research_frontier.py
from __future__ import annotations

from collections.abc import Callable, Mapping
from typing import Any

from .errors import ModelResearchExecutionError
from .model_collaboration import MODEL_WORKFLOW_PROPOSAL_SCHEMA_VERSION
from .model_plan_approval import _compute_canonical_json_digest
from .model_research_execution import MODEL_RESEARCH_RECEIPT_SCHEMA_VERSION


MODEL_RESEARCH_FRONTIER_SCHEMA_VERSION = "kyuubiki.model-research-frontier/v2"
ReceiptVerifier = Callable[[Mapping[str, Any]], bool]
FrontierVerifier = Callable[[Mapping[str, Any]], bool]
_SUBMIT_ACTIONS = {"fem_submit", "workflow_submit_catalog", "workflow_submit_graph"}


class ModelFrontierDigestVerifier:
    def __init__(self, expected_digest: str) -> None:
        if not _valid_plan_digest(expected_digest):
            _fail("expected research frontier digest is invalid")
        self.expected_digest = expected_digest

    def __call__(self, current: Mapping[str, Any]) -> bool:
        verify_model_research_frontier_digest(current, self.expected_digest)
        return True


def start_model_research_frontier(
    receipt: Mapping[str, Any], receipt_verifier: ReceiptVerifier
) -> dict[str, Any]:
    _validate_receipt(receipt)
    _verify_receipt(receipt, receipt_verifier)
    record = _last_record(receipt)
    if receipt["status"] == "failed":
        return _blocked_frontier(receipt, record, receipt["plan_digest"], None, 1)
    if record.get("action") not in _SUBMIT_ACTIONS:
        _fail("initial research receipt must end with a supported job submission")
    job = record.get("output")
    job_id = job.get("job", {}).get("job_id") if isinstance(job, Mapping) else None
    if not isinstance(job_id, str) or not job_id.strip():
        _fail("job submission receipt did not contain job.job_id")
    return _frontier(
        receipt,
        record,
        origin_plan_digest=receipt["plan_digest"],
        stage="waiting_for_job",
        job_id=job_id,
        next_action="job_wait",
        transition_count=1,
    )


def compute_model_research_frontier_digest(current: Mapping[str, Any]) -> str:
    validate_model_research_frontier(current)
    return _compute_canonical_json_digest(_frontier_digest_projection(current))


def verify_model_research_frontier_digest(
    current: Mapping[str, Any], expected_digest: str
) -> None:
    if not _valid_plan_digest(expected_digest):
        _fail("expected research frontier digest is invalid")
    if compute_model_research_frontier_digest(current) != expected_digest:
        _fail("persisted research frontier digest does not match trusted state")


def advance_model_research_frontier(
    current: Mapping[str, Any],
    receipt: Mapping[str, Any],
    frontier_verifier: FrontierVerifier,
    receipt_verifier: ReceiptVerifier,
) -> dict[str, Any]:
    validate_model_research_frontier(current)
    _verify_frontier(current, frontier_verifier)
    _validate_receipt(receipt)
    _verify_receipt(receipt, receipt_verifier)
    if (
        receipt["session_id"] != current["session_id"]
        or receipt["workflow_id"] != current["workflow_id"]
    ):
        _fail("research receipt does not match frontier session and workflow")
    expected = current.get("next_action")
    if not isinstance(expected, str) or not expected:
        _fail("research frontier has no executable next action")
    record = _last_record(receipt)
    if receipt["status"] == "failed":
        return _blocked_frontier(
            receipt,
            record,
            current["origin_plan_digest"],
            current.get("job_id"),
            current["transition_count"] + 1,
        )
    if record.get("action") != expected:
        _fail(
            f"research receipt ended with {record.get('action')}; "
            f"frontier requires {expected}"
        )
    if record.get("job_id") != current.get("job_id"):
        _fail("research receipt job_id does not match frontier binding")

    if expected == "job_wait":
        return _advance_wait(current, receipt, record)
    if expected == "result_fetch":
        return _frontier(
            receipt,
            record,
            origin_plan_digest=current["origin_plan_digest"],
            stage="ready_to_validate",
            job_id=current["job_id"],
            next_action=None,
            transition_count=current["transition_count"] + 1,
        )
    _fail(f"unsupported frontier next action: {expected}")


def build_model_research_frontier_proposal(
    current: Mapping[str, Any],
    frontier_verifier: FrontierVerifier,
) -> dict[str, Any]:
    validate_model_research_frontier(current)
    _verify_frontier(current, frontier_verifier)
    action = current.get("next_action")
    if not isinstance(action, str) or not action:
        _fail("research frontier has no executable next action")
    job_id = current.get("job_id")
    if not isinstance(job_id, str) or not job_id:
        _fail("research frontier has no bound job_id")
    return {
        "schema_version": MODEL_WORKFLOW_PROPOSAL_SCHEMA_VERSION,
        "session_id": current["session_id"],
        "summary": f"Advance verified research frontier with {action}",
        "calls": [
            {
                "id": f"frontier-{current['transition_count'] + 1}-{action}",
                "action": action,
                "payload": {"job_id": job_id},
                "reason": "Use the job identifier retained from verified execution evidence",
            }
        ],
    }


def _advance_wait(
    current: Mapping[str, Any],
    receipt: Mapping[str, Any],
    record: Mapping[str, Any],
) -> dict[str, Any]:
    output = record.get("output")
    terminal = output.get("terminal") if isinstance(output, Mapping) else None
    job = terminal.get("job") if isinstance(terminal, Mapping) else None
    status = job.get("status") if isinstance(job, Mapping) else None
    if status == "completed":
        return _frontier(
            receipt,
            record,
            origin_plan_digest=current["origin_plan_digest"],
            stage="ready_to_fetch_result",
            job_id=current["job_id"],
            next_action="result_fetch",
            transition_count=current["transition_count"] + 1,
            job_status=status,
        )
    if status in {"failed", "cancelled"}:
        return _frontier(
            receipt,
            record,
            origin_plan_digest=current["origin_plan_digest"],
            stage="blocked",
            job_id=current["job_id"],
            next_action=None,
            transition_count=current["transition_count"] + 1,
            job_status=status,
            blocking_reason=f"job reached terminal status {status}",
        )
    if not isinstance(status, str):
        _fail("job_wait receipt did not contain terminal.job.status")
    _fail(f"job_wait returned non-terminal status {status}")


def _frontier(
    receipt: Mapping[str, Any],
    record: Mapping[str, Any],
    *,
    origin_plan_digest: str,
    stage: str,
    job_id: str | None,
    next_action: str | None,
    transition_count: int,
    job_status: str | None = None,
    blocking_reason: str | None = None,
) -> dict[str, Any]:
    return {
        "schema_version": MODEL_RESEARCH_FRONTIER_SCHEMA_VERSION,
        "session_id": receipt["session_id"],
        "workflow_id": receipt["workflow_id"],
        "origin_plan_digest": origin_plan_digest,
        "stage": stage,
        "job_id": job_id,
        "next_action": next_action,
        "transition_count": transition_count,
        "evidence": {
            "approval_id": receipt.get("approval_id"),
            "plan_digest": receipt["plan_digest"],
            "action": record.get("action"),
            "record_index": record.get("index"),
            "authority": record.get("authority"),
            "job_status": job_status,
        },
        "blocking_reason": blocking_reason,
    }


def _blocked_frontier(
    receipt: Mapping[str, Any],
    record: Mapping[str, Any],
    origin_plan_digest: str,
    job_id: str | None,
    transition_count: int,
) -> dict[str, Any]:
    reason = record.get("error")
    return _frontier(
        receipt,
        record,
        origin_plan_digest=origin_plan_digest,
        stage="blocked",
        job_id=job_id,
        next_action=None,
        transition_count=transition_count,
        blocking_reason=reason if isinstance(reason, str) else "research execution failed",
    )


def _validate_receipt(receipt: Mapping[str, Any]) -> None:
    if not isinstance(receipt, Mapping):
        _fail("research execution receipt must be a JSON object")
    if (
        receipt.get("schema_version") != MODEL_RESEARCH_RECEIPT_SCHEMA_VERSION
        or receipt.get("execution_authority") != "kyuubiki-headless-sdk"
    ):
        _fail("unsupported or untrusted research execution receipt")
    if not all(
        isinstance(receipt.get(key), str) and receipt[key].strip()
        for key in ("session_id", "workflow_id")
    ) or not _valid_plan_digest(receipt.get("plan_digest")) or not isinstance(
        receipt.get("records"), list
    ) or not receipt["records"]:
        _fail("research execution receipt is incomplete")
    if receipt.get("status") not in {"completed", "failed"}:
        _fail("research execution receipt has an unsupported status")
    final_record = receipt["records"][-1]
    if not isinstance(final_record, Mapping):
        _fail("research execution receipt has an invalid final record")
    if receipt["status"] == "completed" and (
        final_record.get("error") is not None
        or final_record.get("output") is None
        or not isinstance(final_record.get("authority"), str)
    ):
        _fail("completed research receipt has an invalid final record")
    if receipt["status"] == "failed" and not isinstance(final_record.get("error"), str):
        _fail("failed research receipt has no final error")


def validate_model_research_frontier(current: Mapping[str, Any]) -> None:
    if not isinstance(current, Mapping):
        _fail("research frontier must be a JSON object")
    if (
        current.get("schema_version") != MODEL_RESEARCH_FRONTIER_SCHEMA_VERSION
        or not isinstance(current.get("session_id"), str)
        or not current["session_id"].strip()
        or not isinstance(current.get("workflow_id"), str)
        or not current["workflow_id"].strip()
        or not _valid_plan_digest(current.get("origin_plan_digest"))
        or not isinstance(current.get("evidence"), Mapping)
        or not _valid_plan_digest(current["evidence"].get("plan_digest"))
        or not _valid_evidence(current["evidence"])
        or not isinstance(current.get("transition_count"), int)
        or isinstance(current["transition_count"], bool)
        or current["transition_count"] < 1
    ):
        _fail("research frontier is incomplete or uses an unsupported schema")
    job_id = current.get("job_id")
    has_job = isinstance(job_id, str) and bool(job_id.strip())
    stage = current.get("stage")
    next_action = current.get("next_action")
    blocking_reason = current.get("blocking_reason")
    valid_state = (
        stage == "waiting_for_job"
        and has_job
        and next_action == "job_wait"
        and blocking_reason is None
    ) or (
        stage == "ready_to_fetch_result"
        and has_job
        and next_action == "result_fetch"
        and blocking_reason is None
    ) or (
        stage == "ready_to_validate"
        and has_job
        and next_action is None
        and blocking_reason is None
    ) or (
        stage == "blocked"
        and next_action is None
        and isinstance(blocking_reason, str)
        and bool(blocking_reason.strip())
    )
    if not valid_state:
        _fail("research frontier stage and next action are inconsistent")


def _verify_receipt(receipt: Mapping[str, Any], verifier: ReceiptVerifier) -> None:
    if not callable(verifier) or not verifier(receipt):
        _fail("caller receipt verifier rejected research execution receipt")


def _valid_plan_digest(value: Any) -> bool:
    return (
        isinstance(value, str)
        and len(value) == 71
        and value.startswith("sha256:")
        and all(character in "0123456789abcdef" for character in value[7:])
    )


def _valid_evidence(evidence: Mapping[str, Any]) -> bool:
    action = evidence.get("action")
    record_index = evidence.get("record_index")
    approval_id = evidence.get("approval_id")
    authority = evidence.get("authority")
    return (
        isinstance(action, str)
        and bool(action)
        and action[0].islower()
        and action[0].isascii()
        and all(
            character.isascii()
            and (character.islower() or character.isdigit() or character == "_")
            for character in action
        )
        and isinstance(record_index, int)
        and not isinstance(record_index, bool)
        and record_index > 0
        and (approval_id is None or isinstance(approval_id, str))
        and (authority is None or isinstance(authority, str))
        and evidence.get("job_status") in {None, "completed", "failed", "cancelled"}
    )


def _frontier_digest_projection(current: Mapping[str, Any]) -> dict[str, Any]:
    evidence = current["evidence"]
    return {
        "schema_version": current["schema_version"],
        "session_id": current["session_id"],
        "workflow_id": current["workflow_id"],
        "origin_plan_digest": current["origin_plan_digest"],
        "stage": current["stage"],
        "job_id": current.get("job_id"),
        "next_action": current.get("next_action"),
        "transition_count": current["transition_count"],
        "evidence": {
            "approval_id": evidence.get("approval_id"),
            "plan_digest": evidence["plan_digest"],
            "action": evidence.get("action"),
            "record_index": evidence.get("record_index"),
            "authority": evidence.get("authority"),
            "job_status": evidence.get("job_status"),
        },
        "blocking_reason": current.get("blocking_reason"),
    }


def _verify_frontier(current: Mapping[str, Any], verifier: FrontierVerifier) -> None:
    if not callable(verifier) or not verifier(current):
        _fail("caller frontier verifier rejected research frontier")


def _last_record(receipt: Mapping[str, Any]) -> Mapping[str, Any]:
    record = receipt["records"][-1]
    if not isinstance(record, Mapping):
        _fail("research execution receipt has an invalid final record")
    return record


def _fail(message: str) -> None:
    raise ModelResearchExecutionError([message])
